providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,962 @@
+        EVP_MD_CTX_free(ctx->mdctx);
+        EVP_MD_free(ctx->md);
+        ctx->md = md;
+        ctx->mdsize = EVP_MD_size(md);
+        ctx->mdctx = EVP_MD_CTX_new();
+        if (ctx->mdctx == NULL)
+            return 0;
+
+        if (!EVP_DigestInit_ex(ctx->mdctx, md, NULL))
+            return 0;
+    }
+
+    if (ctx->mgf1_mdname[0] != '\0') {
+        EVP_MD *md = EVP_MD_fetch(ctx->libctx, ctx->mgf1_mdname,
+                                  ctx->mgf1_mdprops);
+
+        if (md == NULL)
+            return 0;
+        EVP_MD_free(ctx->mgf1_md);
+        ctx->mgf1_md = md;
+    }
+
+    return 1;
+}
+
+static int setup_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        return 1;
+    if ((ctx->tbuf = OPENSSL_malloc(RSA_size(ctx->rsa))) == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+}
+
+static void clean_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_cleanse(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static void free_tbuf(PROV_RSA_CTX *ctx)
+{
+    if (ctx->tbuf != NULL)
+        OPENSSL_clear_free(ctx->tbuf, RSA_size(ctx->rsa));
+}
+
+static int rsa_sign(void *vprsactx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
+    int ret;
+
+    if (sig == NULL) {
+        *siglen = RSA_size(prsactx->rsa);
+        return 1;
+    }
+
+    if (sigsize < (size_t)RSA_size(prsactx->rsa))
+        return 0;
+
+    if (!setup_md(prsactx))
+        return 0;
+
+    if (prsactx->md != NULL) {
+        if (tbslen != prsactx->mdsize) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST_LENGTH);
+            return 0;
+        }
+
+        if (EVP_MD_type(prsactx->md) == NID_mdc2) {
+            unsigned int sltmp;
+
+            if (prsactx->pad_mode != RSA_PKCS1_PADDING) {
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_PADDING_MODE,
+                               "only PKCS#1 padding supported with MDC2");
+                return 0;
+            }
+            ret = RSA_sign_ASN1_OCTET_STRING(0, tbs, tbslen, sig, &sltmp,
+                                             prsactx->rsa);
+
+            if (ret <= 0) {
+                ERR_raise(ERR_LIB_PROV, ERR_LIB_RSA);
+                return 0;
+            }
+            ret = sltmp;
+        } else if (prsactx->pad_mode == RSA_X931_PADDING) {
+            if ((size_t)RSA_size(prsactx->rsa) < tbslen + 1) {
+                ERR_raise(ERR_LIB_PROV, PROV_R_KEY_SIZE_TOO_SMALL);
+                return 0;
+            }
+            if (!setup_tbuf(prsactx)) {
+                ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+                return 0;
+            }
+            memcpy(prsactx->tbuf, tbs, tbslen);
+            prsactx->tbuf[tbslen] = RSA_X931_hash_id(EVP_MD_type(prsactx->md));
+            ret = RSA_private_encrypt(tbslen + 1, prsactx->tbuf,
```

Comment:
RSA_private_encrypt() currently does
{
    return rsa->meth->rsa_priv_enc
}
I assume this is another PR?

Reply:
Correct. This PR does *not* rewrite the RSA subsystem

Reply:
Also, I don't see the issue with going through rsa->meth per se, as long as we know that we have control over what low-level method we use.

Reply:
As long as the ->meth are not settable in FIPS_MODE its ok..
It seems strange to have a dispatch function that calls a method table..

Reply:
The ->meth should be settable from within FIPS, but only to *our* RSA methods. 

You haven't noticed yet that we have layers upon layers of method structures? However, one might argue that those at such a low level should be thrown away, but that needs to be investigated. Meanwhile, as long as we have control, we should be fine.

Reply:
Note that since there's no engine support in the FIPS module, and no other way to specify a different default RSA method, rsa->meth will point at rsa_pkcs1_ossl_meth (the method structure found in `crypto/rsa/rsa_ossl.c`